alternator/streams.cc

```diff
@@ -0,0 +1,184 @@
+
+template<typename ValueType, typename T>
+struct from_string_helper {
+    static bool Is(const ValueType& v) {
+        try {
+            Get(v);
+            return true;
+        } catch (...) {
+            return false;
+        }
+    }
+    static T Get(const ValueType& v) {
+        auto s = rjson::to_string_view(v);
+        if constexpr (std::is_constructible_v<T, std::string_view>) {
+            return T(s);
+        } else if constexpr (std::is_constructible_v<T, sstring>) {
+            return T(sstring(s));
+        } else {
+            return boost::lexical_cast<T>(s.data(), s.size());
+        }
+    }
+    static ValueType& Set(ValueType&, T) = delete;
+    static ValueType& Set(ValueType& v, T data, typename ValueType::AllocatorType& a) {
+        std::ostringstream ss;
+        ss << data;
+#if __cplusplus >= 202002L
+        auto s = ss.view();
+#else
+        auto s = ss.str();
+#endif
+        return v.SetString(s.data(), s.size(), a);
+    }
+};
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, utils::UUID>
+    : public from_string_helper<ValueType, utils::UUID>
+{};
+
+namespace alternator {
+
+// stream arn _has_ to be 37 or more characters long. ugh...
+class stream_arn : public utils::UUID {
+public:
+    using UUID = utils::UUID;
+    static constexpr char marker = 'S';
+
+    stream_arn() = default;
+    stream_arn(const UUID& uuid)
+        : UUID(uuid)
+    {}
+    stream_arn(std::string_view v)
+        : UUID(v.substr(1))
+    {
+        if (v[0] != marker) {
+            throw std::invalid_argument(std::string(v));
+        }
+    }
+    friend std::ostream& operator<<(std::ostream& os, const stream_arn& arn) {
+        const UUID& uuid = arn;
+        return os << marker << uuid;
+    }
+    friend std::istream& operator>>(std::istream& is, stream_arn& arn) {
+        std::string s;
+        is >> s;
+        arn = stream_arn(s);
+        return is;
+    }
+};
+
+}
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, alternator::stream_arn>
+    : public from_string_helper<ValueType, alternator::stream_arn>
+{};
+
+namespace alternator {
+
+future<alternator::executor::request_return_type> alternator::executor::list_streams(client_state& client_state, service_permit permit, rjson::value request) {
+    _stats.api_operations.list_streams++;
+
+    auto streams_start = rjson::get_opt<stream_arn>(request, "ExclusiveStartStreamArn");
+    auto limit = rjson::get_opt<int>(request, "Limit").value_or(std::numeric_limits<int>::max());
+    auto table = get_table(_proxy, request, false);
+
+    auto& db = _proxy.get_db().local();
+    auto& cfs = db.get_column_families();
+    auto i = cfs.begin();
+    auto e = cfs.end();
+
+    // TODO: the unordered_map here is not really well suited for partial
+    // querying - we're sorting on local hash order, and creating a table
+    // between queries may or may not miss info. But that should be rate,
```

Comment:
There is not a bug. But we could skip data if the set of cf:s is rehashed between calls. I considered sorting, but we have nothing to sort on, because table UUID is based on names, and there are no timestamps in CF:s for creation.